### ISSUE TITLE
Reproduction Problem on XSUM

### ISSUE THREAD

**Issue by hzhwcmhf** [kongds/MIST#1]
Hi, thanks for the interesting work!

I am reproducing your results on XSUM but it is much worse than the reported one in the paper.

My Result: R-1 ~30 / R-2 ~7.5 / R-L ~24
Paper Result: R-1 34.63 / R2 11.29 / R-L 28.70

I download unilm1.2 and train the model by

```
TRAIN_FILE=/your/path/to/train.json
VALID_FILE=/your/path/to/valid.json
OUTPUT_DIR=/your/path/to/save_checkpoints
CACHE_DIR=/your/path/to/transformer_package_cache

MODEL_PATH=/path/to/unilm1.2-base-uncased

# squadqg 30005 steps
# squadqg 50005 steps
# xsum 600005 steps
STEPS=600005

python -m torch.distributed.launch --nproc_per_node=4 train.py\
  --dataset xsum
  --train_file $TRAIN_FILE\
  --valid_file $VALID_FILE\
  --output_dir $OUTPUT_PATH\
  --model_type nat --model_name_or_path $MODEL_PATH\
  --do_lower_case --max_source_seq_length 464 --max_target_seq_length 48\
  --per_gpu_train_batch_size 16 --gradient_accumulation_steps 1\
  --learning_rate 3e-5 --num_warmup_steps 500 --num_training_steps $STEPS\
  --cache_dir $CACHE_DIR\
  --log_dir ${OUTPUT_PATH}/log\
  --keep_prob 0.0\
  --random_prob 0.0\
  --use_glat\
  --tqdm_miniters 100\
  --cotrain_put_target_in_source\ 
  --cotrain_put_target_in_source_same_bert\ 
  --wandb\ # logging with wandb
  --fp16\
  --fp16_opt_level O2
```

I also find the valid performance do not improve about 20k training steps. Can you provide any advice?

**Comment by kongds** [comment]
Hello, Can you report the test result?
By adding follow flags
```
    --valid_only\
    --test_file DATA_PATH/test.json\
    --valid_model_path MODEL_PATH/model_state_dict.bin\
    --test

```

**Comment by hzhwcmhf** [comment]
Training Curve (up to 180k steps, very different from Figure2 in your paper):

![curve](https://user-images.githubusercontent.com/1344510/187824667-1870fbbb-ba0f-408a-b9e6-eda5368242e4.png)

Output file (test set):

[out_f_merge_duplicate.txt](https://github.com/kongds/MIST/files/9466567/out_f_merge_duplicate.txt)

My evaluation result (calling files2rouge)

```
Preparing documents... 0 line(s) ignored
Running ROUGE...
---------------------------------------------
1 ROUGE-1 Average_R: 0.45805 (95%-conf.int. 0.45487 - 0.46150)
1 ROUGE-1 Average_P: 0.25119 (95%-conf.int. 0.24900 - 0.25339)
1 ROUGE-1 Average_F: 0.31703 (95%-conf.int. 0.31456 - 0.31943)
---------------------------------------------
1 ROUGE-2 Average_R: 0.11536 (95%-conf.int. 0.11285 - 0.11781)
1 ROUGE-2 Average_P: 0.06534 (95%-conf.int. 0.06368 - 0.06700)
1 ROUGE-2 Average_F: 0.08141 (95%-conf.int. 0.07954 - 0.08336)
---------------------------------------------
1 ROUGE-L Average_R: 0.36747 (95%-conf.int. 0.36451 - 0.37048)
1 ROUGE-L Average_P: 0.20271 (95%-conf.int. 0.20062 - 0.20476)
1 ROUGE-L Average_F: 0.25507 (95%-conf.int. 0.25287 - 0.25746)
```

**Comment by kongds** [comment]
Is the training data from our download link?（https://github.com/kongds/MIST#download-datasets）

**Comment by hzhwcmhf** [comment]
> Is the training data from our download link?（[https://github.com/kongds/MIST#download-datasets）](https://github.com/kongds/MIST#download-datasets%EF%BC%89)

Yes

**Comment by kongds** [comment]
Most of the training data is on the company's server, but I can't access it due to checkout. (I'll contact my mentor afterwards to find the tensorboard. And current I don't have enough resource to reproduce this experiment)

Also I have some of the training data here on wandb, but it seems that only the results of bert-base-uncased on xsum (https://wandb.ai/royokong/unilm-nat/runs/1185dicy?workspace=user-royokong).

**Comment by kongds** [comment]
In addition, I find that [out_f_merge_duplicate.txt](https://github.com/kongds/MIST/files/9466567/out_f_merge_duplicate.txt) seems not been processed by https://github.com/kongds/MIST/blob/1792971b052b25df5c03d394c276c34900214ad7/train.py#L166-L172

**Comment by hzhwcmhf** [comment]
I forgot to add ``--dataset xsum`` in inference, but it does not make much differences.

```
Running ROUGE...
---------------------------------------------
1 ROUGE-1 Average_R: 0.45800 (95%-conf.int. 0.45481 - 0.46149)
1 ROUGE-1 Average_P: 0.25116 (95%-conf.int. 0.24896 - 0.25337)
1 ROUGE-1 Average_F: 0.31700 (95%-conf.int. 0.31453 - 0.31940)
---------------------------------------------
1 ROUGE-2 Average_R: 0.11538 (95%-conf.int. 0.11287 - 0.11783)
1 ROUGE-2 Average_P: 0.06533 (95%-conf.int. 0.06369 - 0.06701)
1 ROUGE-2 Average_F: 0.08142 (95%-conf.int. 0.07952 - 0.08336)
---------------------------------------------
1 ROUGE-L Average_R: 0.36710 (95%-conf.int. 0.36419 - 0.37010)
1 ROUGE-L Average_P: 0.20252 (95%-conf.int. 0.20044 - 0.20461)
1 ROUGE-L Average_F: 0.25483 (95%-conf.int. 0.25262 - 0.25726)
```

**Comment by kongds** [comment]
Hello
I checked the code and find the different training curve seems caused by removing repeat words before calculating Rouge in current code. (This may affect the checkpoint selection)
https://github.com/kongds/MIST/blob/1792971b052b25df5c03d394c276c34900214ad7/train.py#L164
Can you try to remove this? (Our results is from checkpoint at 575000 training step)

**Comment by hzhwcmhf** [comment]
@kongds Yes, I can reproduce your result now. I train the model with the line removed, and then inference with the line added.

I think there are also some random factors, where ROUGE-L of my best trial starts to climb up at 150k steps, but some other trials do not have any sign of improvement even at 200k.

**Comment by kongds** [comment]
Thanks for reproduction.

For random factors, what is the difference between different trials ? (random seeds?)

**Comment by hzhwcmhf** [comment]
I didn't change anything in the script except the checkpoint path. Maybe there are some other random issues like undeterministic operations in pytorch.
Since our devices are running other models, I am not going to dig deeper. Just mention it if there are others trying to reproduce the results.

**Comment by kongds** [comment]
Got it. Thanks again.
